### PR TITLE
Add -j flag for multi-threaded validation

### PR DESCRIPTION
**Pull request by simonlindholm**
Only parallelizes test case execution and submission compilation at the moment; input validation could be added with a bit of additional work, but it felt less important to me since it takes almost no time with validators written in C++ as we usually have it.

mypy is happy with the refactoring and I've done some lightweight testing, but it's possible that there's some breakage along less happy paths, though probably restricted to -j.

### CODE REVIEW

File: problemtools/run/__init__.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def find_programs(path, pattern='.*', language_config=None, work_dir=None,
 
 
 def get_program(path, language_config=None, work_dir=None, include_dir=None,
-                allow_validation_script=False):
+                allow_validation_script=False) -> Program|None:
```
Comment (Tagl) [inline problemtools/run/__init__.py:65]: Perhaps use
```suggestion
                allow_validation_script=False) -> Optional[Program]:
```
instead?

File: problemtools/run/buildrun.py [hunk 2]
```diff
@@ -53,32 +55,27 @@ def __init__(self, path, work_dir=None, include_dir=None):
...
 
-    def __str__(self):
+    def __str__(self) -> str:
         """String representation"""
         return '%s/' % (self.path)
 
 
-    _compile_result = None
-    def compile(self):
+    def do_compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/buildrun.py:63]:
```suggestion
    def do_compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/run/checktestdata.py [hunk 3]
```diff
@@ -22,27 +22,23 @@ def __init__(self, path):
...
 
-    def __str__(self):
+    def __str__(self) -> str:
         """String representation"""
         return '%s' % (self.args[0])
 
 
-    _compile_result = None
-    def compile(self):
+    def do_compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/checktestdata.py:33]:
```suggestion
    def do_compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/run/program.py [hunk 4]
```diff
@@ -50,12 +55,23 @@ def run(self, infile='/dev/null', outfile='/dev/null', errfile='/dev/null',
 
         return status, runtime
 
-    def code_size(self):
+    def compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/program.py:58]:
```suggestion
    def compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/run/program.py [hunk 5]
```diff
@@ -50,12 +55,23 @@ def run(self, infile='/dev/null', outfile='/dev/null', errfile='/dev/null',
...
+        with self._compile_lock:
+            if self._compile_result is None:
+                self._compile_result = self.do_compile()
+            return self._compile_result
+
+    def do_compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/program.py:64]:
```suggestion
    def do_compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/run/source.py [hunk 6]
```diff
@@ -80,24 +81,18 @@ def __init__(self, path, language, work_dir=None, include_dir=None):
...
 
-    def code_size(self):
+    def code_size(self) -> int:
         return sum(os.path.getsize(x) for x in self.src)
 
 
-    _compile_result = None
-
-    def compile(self):
+    def do_compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/source.py:88]:
```suggestion
    def do_compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/run/viva.py [hunk 7]
```diff
@@ -22,26 +22,22 @@ def __init__(self, path):
...
 
 
     def __str__(self):
         """String representation"""
         return '%s' % (self.args[0])
 
 
-    _compile_result = None
-    def compile(self):
+    def do_compile(self) -> tuple[bool, str|None]:
```
Comment (Tagl) [inline problemtools/run/viva.py:33]:
```suggestion
    def do_compile(self) -> tuple[bool, Optional[str]]:
```

File: problemtools/verifyproblem.py [hunk 8]
```diff
@@ -87,6 +91,15 @@ class VerifyError(Exception):
     pass
 
 
+class Context:
+    def __init__(self, args: argparse.Namespace, executor: ThreadPoolExecutor|None) -> None:
```
Comment (Tagl) [inline problemtools/verifyproblem.py:95]:
```suggestion
    def __init__(self, args: argparse.Namespace, executor: Optional[ThreadPoolExecutor]) -> None:
```

File: problemtools/verifyproblem.py [hunk 9]
```diff
@@ -54,7 +58,7 @@ def __init__(self, verdict: str, score: float|None=None, reason: str|None=None,
         self.runtime_testcase: TestCase|None = None
         self.runtime = -1.0
         self.ac_runtime = -1.0
-        self.ac_runtime_testcase = None
+        self.ac_runtime_testcase: TestCase|None = None
```
Comment (Tagl) [inline problemtools/verifyproblem.py:61]:
```suggestion
        self.ac_runtime_testcase: Optional[TestCase] = None
```

File: problemtools/verifyproblem.py [hunk 10]
```diff
@@ -87,6 +91,15 @@ class VerifyError(Exception):
     pass
 
 
+class Context:
+    def __init__(self, args: argparse.Namespace, executor: ThreadPoolExecutor|None) -> None:
+        self.data_filter: Pattern[str] = args.data_filter
+        self.submission_filter: Pattern[str] = args.submission_filter
+        self.fixed_timelim: int|None = args.fixed_timelim
```
Comment (Tagl) [inline problemtools/verifyproblem.py:98]:
```suggestion
        self.fixed_timelim: Optional[int] = args.fixed_timelim
```

File: problemtools/verifyproblem.py [hunk 11]
```diff
@@ -422,7 +424,7 @@ def get_subgroups(self) -> list[TestCaseGroup]:
         return [child for child in self._items if isinstance(child, TestCaseGroup)]
 
 
-    def get_subgroup(self, name):
+    def get_subgroup(self, name: str) -> TestCaseGroup|None:
```
Comment (Tagl) [inline problemtools/verifyproblem.py:427]:
```suggestion
    def get_subgroup(self, name: str) -> Optional[TestCaseGroup]:
```

File: problemtools/verifyproblem.py [hunk 12]
```diff
@@ -706,12 +711,12 @@ def __init__(self, problem: Problem):
     def __str__(self) -> str:
         return 'problem configuration'
 
-    def get(self, key: str|None=None):
+    def get(self, key: str|None=None) -> Any:
```
Comment (Tagl) [inline problemtools/verifyproblem.py:714]:
```suggestion
    def get(self, key: Optional[str]=None) -> Any:
```

File: problemtools/verifyproblem.py [hunk 13]
```diff
@@ -1643,6 +1670,114 @@ def validate(self, testcase: TestCase, submission_output: str) -> SubmissionResu
...
+                return [item]
+        elif item not in self._done_groups:
+            ret = []
+            for child in item.get_testcases() + item.get_subgroups():
+                ret.extend(self._gather_testcases(child))
+            return ret
+        else:
+            return []
+
+    def _next_job(self) -> TestCase|None:
```
Comment (Tagl) [inline problemtools/verifyproblem.py:1759]:
```suggestion
    def _next_job(self) -> Optional[TestCase]:
```

File: problemtools/run/__init__.py (hunk 1 above)
Comment (simonlindholm) [reply problemtools/run/__init__.py:65]: I think `|None` is the more modern version (added in 3.10)? It matches what the rest of the code base uses, too. (I do agree personally that `|None` is a bit ugly.)

**Comment by simonlindholm** [comment]
> Good suggestion! Why the letter j? Just because t is in use?

-j is the customary flag for threading. Also used by e.g. `make`, and by `generatedata.py` in this repo.

File: problemtools/run/__init__.py (hunk 1 above)
Comment (Tagl) [reply problemtools/run/__init__.py:65]: I was unaware of that and have now read through the following were this is debated: https://discuss.python.org/t/clarification-for-pep-604-is-foo-int-none-to-replace-all-use-of-foo-optional-int/26945/44

You can ignore all my suggestions then, I have no strong preference on this at all.

**Review by pehrsoderman — APPROVED** [review]

**Pull request merged**